### ISSUE TITLE
AGE study

### ISSUE THREAD

**Issue by AlexLacson** [LindsayXX/NEM#2]
- [x] Study network architecture

- [x] Training procedure

- [x] Gradients and backprop procedure

- [x] Parameter settings

- [x] Optimization methods

- [x] Data preprocess/augmentation

**Comment by LindsayXX** [comment]
dcgan64px.py build the encoder and generator (feed-forward) model that contains 5 conv layers and 5 conv transpose layers respectively. The encoder mapping the data space(size of input channel(opt.nc) * 64 *64) all the way to the latent space(size of latent space opt.nz * 1 * 1). The encoder and generator in dcgan32px.py are slightly different.

**Comment by AlexLacson** [comment]
**Training procedure:**

1. Update encoder weights

- Mini batch of real images input
-  Feed through encoder convolutional network, output M-dim projected on sphere
- Fit output to Gaussian mean and var, compute loss KL with N(0,1), which is to be **minimized**

- Sample Z~Uniform on M-dim sphere (normalized N(0,1))
- Feed through decoder/generator transpose convolution, output fake image
- Feed through encoder, output M-dim on sphere
- Fit mean and var, compute loss KL with N(0, 1), which is to be **maximized**

- Sum losses, run backpropagation, update weights

2. Update decoder/generator weights

- Sample Z~Uniform on M-dim sphere (normalized N(0,1))
- Feed through decoder/generator transpose convolution, output fake image
- Feed through encoder, output M-dim on sphere
- Fit mean and var, compute loss KL with N(0, 1), which is to be **minimized**
- Compute regularization term L2 dist between z and e[g(z)] 

- Sum losses, run backprop, update weights

3. 

- Alternate between updating encoder and generator, with two updates of the generator per update of encoder
